### ISSUE TITLE
Mounted device rename problem with rsync

### ISSUE THREAD

**Issue by michal7918** [andyholmes/gnome-shell-extension-gsconnect#453]
**Is your feature request related to a problem? Please describe.**
I'm trying to rsync files from desktop to mounted phone filesystem. Transfer of files works very well, but updating is broken. The reason is that rsync first transfers updated file with different filename and then do rename it on destination. Anyway rename file and overwrite existing operation is not permitted by SSHFS filesystem, resulting in error. Rsync's --inplace option doesn't seem to fix the issue.

**Describe the solution you'd like**
According to various documentation, it should be possible to mount sshs filesystem with workaround=rename option which should fix this issue.  

**System Details (please complete the following information):**
 - GSConnect version: 19 
   - Installed from: [e.g. extensions.gnome.org]

**Comment by andyholmes** [comment]
In the next stable release (v21) `sshfs` support will be dropped in favour of GVFS, although I haven't tested if renaming will work with `rsync`.

If you want to implement this for your own purposes (`<=v20`) though, you can change the `sshfs` options in `~/.local/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/plugins/sftp.js`:

https://github.com/andyholmes/gnome-shell-extension-gsconnect/blob/40061377ac8e8e0a95162cd4c0a45abc4f1900bb/src/service/plugins/sftp.js#L249-L277

Probably you can just add the line `'-o', 'workaround=rename'` to the list of arguments and kill the server to make it restart.

    $ touch ~/.local/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/daemon.js

**Comment by michal7918** [comment]
Excellent, your suggestion worked for me, thank you for pointing me to right part of the code!

**Comment by ferdnyc** [comment]
Thing is, though, that's going away when `sshfs` support is dropped, and there's no way to set options for the ssh mounts created by Gvfs.

Running `rsync` over an sshfs mount is kind of a bad idea just generally — it's the sort of thing you want to avoid if at all possible (unfortunately it probably isn't, in this case). Basically everything about the way it functions is working **against** you, when you're syncing to a network filesystem, but the local `rsync` process has to run both sides of the connection.

Just for starters, you're better off just blindly transferring the _entire_ file anytime there's an update, because any sort of checksumming to determine what parts, if any, have changed in any of the files is actually **less** efficient. (In order to compute the checksum, the remote version of the file has to first be downloaded to the local computer! Which right away destroys all of the efficiency that might be gained by doing checksumming.)

Because of this — because the most efficient method of running rsync locally to a remote mount is just to blindly push all updates — you can actually run `rsync` with the flags `-u -W --delete-before`, to push updates without needing to enable renaming. The destination file on the remote end will just be deleted outright, before `rsync` transfers an updated copy from the local machine. 

(The `-u` flag tells `rsync` to compare the modification times and sizes of the local and remote files, to determine what needs to be updated. Only files that are newer on the source than the destination — meaning, they've been updated since the last `rsync` run — will be transferred. The `-W` flag says "whole files only", and avoids using the delta algorithm, because of the aforementioned inefficiency in doing comparisons/checksums.)

**TL;DR: `rsync -u -W --delete-before ...` should work without needing rename support, and is probably as efficient as you can get with one-sided rsync**

In practice, at least in my own testing, this still... isn't really great. Transfers are dog-slow, and the mount for my phone's filesystem has a nasty habit of just suddenly dropping right in the middle of a transfer, necessitating a visit to the User Menu so I can click "Mount" again, and then a quick check to make sure that the mount point is still the same (port or IP hasn't changed), before I can start `rsync` up again. But at least it will pick back up at the file that was transferring when the mount disappeared — though it has to start that file over from the beginning, again, of course.

**Comment by michal7918** [comment]
Thanks for further elaboration but `--delete-before` is not doing the job for me for some reason. Still I get rename not permitted errors. Anyways thanks for suggesting `-u -W` I assume that really increases efficiency. 

I agree that whole thing is cumbersome. Anyway android device erratic disconnecting can be partly fixed by blacklisting battery optimization for KDEConnect app. I did not notice battery drain penalty form this.

**Comment by ferdnyc** [comment]
> Thanks for further elaboration but `--delete-before` is not doing the job for me for some reason.

Huh, really? That's odd. What's your entire `rsync` command line? I had no problem updating existing dirs from my desktop to my phone over the Gvfs mount using:
```sh
rsync -rltvz -i -u -W --delete-before --progress
```
(Modulo the connection flakiness I mentioned.)

...And I'm just now realizing how incredibly **stupid** it is, that I have the compression flag set in there. So, yeah, definitely drop `-z` as well.

> Anyway android device erratic disconnecting can be partly fixed by blacklisting battery optimization for KDEConnect app.

That's a good suggestion, Samsung's stupid Samsung+ app is pretty aggressive with that stuff, too. (I just hope it respects the app whitelist.)
